### PR TITLE
[DEV-1757] add fee table + discount codes

### PR DESCRIPTION
**Pull request by Yannick1712**
#### Pre-Release

- [ ] Check migrations
  - No database related infos (sqldb-xxx)
  - Impact on GS (new/removed columns)
- [ ] Check for linter errors (in PR)
- [ ] Test basic user operations
  - Login/logout
  - Registration
  - Update user data
  - Create/delete payment routes

#### Post-Release

- Test basic user operations
- Monitor application insights log

### CODE REVIEW

**Comment by Yannick1712** [comment]
Add migration

File: src/subdomains/core/fee/fee.service.ts [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+    });
+
+    if (existing) throw new BadRequestException('Fee already created');
+    if (dto.type !== FeeType.DISCOUNT && dto.createDiscountCode)
+      throw new BadRequestException('Only discount Fees can have a discountCode');
+    if (dto.type === FeeType.BASE && !dto.accountType)
+      throw new BadRequestException('Base Fees must have an accountType');
+
+    // create the entity
+    const buy = this.feeRepo.create(dto);
```
Comment (davidleomay) [inline src/subdomains/core/fee/fee.service.ts:55]: Copy&paste? 😉

File: src/subdomains/core/fee/fee.service.ts [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+  ) {}
+
+  async createFee(dto: CreateFeeDto): Promise<Fee> {
+    // check if exists
+    const existing = await this.feeRepo.findOne({
+      where: {
+        label: dto.label,
+        direction: dto.direction,
+      },
+    });
```
Comment (davidleomay) [inline src/subdomains/core/fee/fee.service.ts:46]: Maybe add a unique index instead?

File: src/subdomains/core/fee/fee.service.ts [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+    );
+  }
+
+  private isExpiredFee(fee: Fee, request: FeeRequest): boolean {
+    return (
+      !fee ||
+      (fee.expiryDate && fee.expiryDate < new Date()) ||
+      (fee.accountType && fee.accountType !== request.accountType)
+    );
+  }
```
Comment (davidleomay) [inline src/subdomains/core/fee/fee.service.ts:224]: Can we just combine these two methods and rename to `isMatchingFee`? Or remove the accountType check on `isExpiredFee`?

File: src/subdomains/core/fee/fee.service.ts [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+  async addDiscountCodeUser(userData: UserData, discountCode: string): Promise<void> {
+    const fee = await this.getFeeByDiscountCode(discountCode);
+    if (this.isExpiredFee(fee, { accountType: userData.accountType }))
+      throw new BadRequestException('Discount code is expired');
+
+    if (fee.maxUsages && (await this.hasMaxUsageExceeded(fee)))
+      throw new BadRequestException('Max usages for discount code taken');
+
+    await this.userDataService.addDiscountCode(userData, fee.id.toString());
+  }
```
Comment (davidleomay) [inline src/subdomains/core/fee/fee.service.ts:87]: Should we really throw errors, or just not add the codes?

File: src/subdomains/core/fee/fee.service.ts [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+
+  async addFeeInternal(userData: UserData, feeId: number): Promise<void> {
+    const fee = await this.feeRepo.findOneBy({ id: feeId });
+    if (this.isExpiredFee(fee, { accountType: userData.accountType }))
+      throw new BadRequestException('Discount code is expired');
+
+    if (fee.maxUsages && (await this.hasMaxUsageExceeded(fee)))
+      throw new BadRequestException('Max usages for discount code taken');
```
Comment (davidleomay) [inline src/subdomains/core/fee/fee.service.ts:95]: This looks the same than lines 80-84. Maybe we can add a method `verifyFee` or something?

File: src/subdomains/core/fee/fee.service.ts [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+    let userDiscount = 0;
+
+    for (const fee of fees) {
+      switch (fee.type) {
+        case FeeType.BASE:
+          if (!userBaseFee || userBaseFee > fee.value) userBaseFee = fee.value;
+          break;
+        case FeeType.CUSTOM:
+          if (!userCustomFee || userCustomFee < fee.value) userCustomFee = fee.value;
+          break;
```
Comment (davidleomay) [inline src/subdomains/core/fee/fee.service.ts:133]: Why do we take the highest custom fee?

File: src/subdomains/core/fee/fee.service.ts [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+    }
+
+    if (!userBaseFee) throw new InternalServerErrorException('Base Fee is missing');
+    if (userBaseFee - userDiscount < 0) {
+      this.logger.warn(`UserDiscount higher userBaseFee! UserDataId: ${userDataId}`);
+      userDiscount = 0;
+    }
+
+    return userCustomFee ?? userBaseFee - userDiscount;
+  }
```
Comment (davidleomay) [inline src/subdomains/core/fee/fee.service.ts:147]: Maybe we could rearrange this?
```ts
private calculateFee(fees: Fee[], userDataId?: number): number {
   const customFee = Math.min(...fees.filter((f) => f.type === FeeType.CUSTOM).map((f) => f.value));
   if (customFee !== Infinity) return customFee;

   const baseFee = Math.min(...fees.filter((f) => f.type === FeeType.BASE).map((f) => f.value));
   const discount = Math.max(...fees.filter((f) => f.type === FeeType.DISCOUNT).map((f) => f.value));
   if (baseFee === Infinity) throw new InternalServerErrorException('Base fee is missing');
   ...
}
```

File: src/subdomains/core/fee/fee.service.ts [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+    }
+
+    return userFees.sort((a, b) => {
+      const TypeOrder: { [b in FeeType]: number } = {
+        [FeeType.CUSTOM]: 1,
+        [FeeType.BASE]: 2,
+        [FeeType.DISCOUNT]: 3,
+      };
+      return TypeOrder[a.type] - TypeOrder[b.type];
+    });
```
Comment (davidleomay) [inline src/subdomains/core/fee/fee.service.ts:180]: Why this sorting?

File: src/subdomains/core/fee/fee.service.ts [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+      },
+      {
+        type: FeeType.DISCOUNT,
+        accountType: IsNull(),
+        discountCode: IsNull(),
+      },
+    ]);
+
+    return discounts.filter((discount) => this.isValidFee(discount, { ...request, accountType: request.accountType }));
+  }
```
Comment (davidleomay) [inline src/subdomains/core/fee/fee.service.ts:207]: Did you forget the custom fees?

File: src/subdomains/core/fee/fee.service.ts [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+
+  async getUserFee(request: UserFeeRequest): Promise<number> {
+    const userFees = await this.getValidUserFees(request);
+
+    return this.calculateFee(userFees, request.userData.id);
+  }
+
+  async getDefaultFee(request: FeeRequestBase, accountType = AccountType.PERSONAL): Promise<number> {
+    const baseFees = await this.getBaseFees({ ...request, accountType });
+    const discounts = await this.getFreeDiscounts({ ...request, accountType });
```
Comment (davidleomay) [inline src/subdomains/core/fee/fee.service.ts:114]: I would try to combine this somehow with `getValidUserFees`

File: src/subdomains/core/fee/fee.service.ts [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+
+    return userFees.sort((a, b) => {
+      const TypeOrder: { [b in FeeType]: number } = {
+        [FeeType.CUSTOM]: 1,
+        [FeeType.BASE]: 2,
+        [FeeType.DISCOUNT]: 3,
+      };
+      return TypeOrder[a.type] - TypeOrder[b.type];
+    });
+  }
```
Comment (davidleomay) [inline src/subdomains/core/fee/fee.service.ts:181]: I would try to 
1. get all matching fees in one single call (e.g. filtering by accountType shouldn't be necessary, this is done in `isValidFee`), incl. getting all the discounts,
2. find all expired discount fees (in this array) and remove the fee from UserData
3. filter all matching fees with the `isValidFee` method

**Review by davidleomay — COMMENTED** [review]
Fix linter warnings

File: src/subdomains/supporting/payment/services/fee.service.ts [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+export interface OptionalFeeRequest extends FeeRequestBase {
+  userData?: UserData;
+  accountType?: AccountType;
+}
+
+export interface FeeRequestBase {
+  direction?: FeeDirectionType;
+  asset?: Asset;
+  txVolume?: number;
+}
```
Comment (davidleomay) [inline src/subdomains/supporting/payment/services/fee.service.ts:32]: `direction` required?

File: src/subdomains/supporting/payment/services/fee.service.ts [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+    return userFees.filter((fee) => this.isValidFee(fee, { ...request, accountType }));
+  }
+
+  private isValidFee(fee: Fee, request: FeeRequest): boolean {
+    return !(
+      this.isExpiredFee(fee) ||
+      (fee.accountType && fee.accountType !== request.accountType) ||
+      (fee.direction && fee.direction !== request.direction) ||
+      (fee.assetList && request.asset && !fee.assetList.includes(request.asset?.id)) ||
+      (fee.maxTxVolume && fee.maxTxVolume < request.txVolume)
```
Comment (davidleomay) [inline src/subdomains/supporting/payment/services/fee.service.ts:165]: Remove `&& request.asset`

Comment (davidleomay) [reply src/subdomains/supporting/payment/services/fee.service.ts:165]: `fee.assetList.length`

**Review by davidleomay — APPROVED** [review]

**Pull request merged**